src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNINpHandle.cs

```diff
@@ -312,8 +322,29 @@ public override uint EnableSsl(uint options)
                 _validateCert = (options & TdsEnums.SNI_SSL_VALIDATE_CERTIFICATE) != 0;
                 try
                 {
-                    _sslStream.AuthenticateAsClient(_targetServer, null, SupportedProtocols, false);
-                    _sslOverTdsStream.FinishHandshake();
+                    if (_isTDS8)
+                    {
+#if NETCOREAPP
+                        SslApplicationProtocol TDS8 = new("tds/8.0");
+
+                        SslClientAuthenticationOptions sslClientOptions = new()
+                        {
+                            TargetHost = _serverNameIndication,
+                            ApplicationProtocols = new List<SslApplicationProtocol>() { TDS8 },
+                            EnabledSslProtocols = SupportedProtocols,
+                            ClientCertificates = null,
+                        };
+                        _sslStream.AuthenticateAsClientAsync(sslClientOptions).Wait();
```

Comment:
This is using an async function and then awaiting it synchronously, if @roji sees you doing this you'll make him sad. See if there is a naturally synchronous `AuthenticateAsClient` method you can use instead and avoid sync-over-async.

Reply:
there is no AuthenticateAsClient that accepts SslClientAuthenticationOption as an argument in netcore 3.1. It started from net5 and after.

Reply:
Netcore 3.1 goes out of support at the end of this year (it's LTS) and net5 is already unsupported. https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-core

You can use compiler ifdefs to use the current pattern only on 3.1 and lower and then use the sync overload from net5 onwards. In this case I think it's acceptable to do that because it will avoid more threading problems.